Fix up nova inventory to be modern

emonty opened the pull request:
Update the nova inventory module to be more like the ec2 and rax modules. Notably:
- added support for nova.ini to express multiple clouds that the inventory should span
- added caching similar to the ec2 module
- aligned authentication and floating ip handling with the other nova modules
- added "--private" flag so that one can specify "please use the private IP (fixed IP) of the instance as ansible_ssh_host

CaptTofu committed:
Complete re-work to work with current Nova, private feature

emonty committed:
Clean up a few things about config parsing

Using ConfigParser defaults more extensively to cut down on the
amount of logic in config parsing. Also, remove auth_system - it's
a deprecated/removed feature upstream.

emonty committed:
Align default values with other openstack modules

The default username elsewhere is 'admin' and the default project_id
is the same as the username if not set.

emonty committed:
Add flavor_name grouping

emonty committed:
Handle the case of no nova.ini file existing

The nova inventory modue will happily work with no config file,
except for how we're reading default values.

emonty committed:
Support multiple OpenStack clouds

It's entirely regular in OpenStack to have multiple cloud endpoints
that an infrastructure spans. Add support to nova.ini for expressing
that, and additionally add a few more groups and hostvars so that,
for instance, it would be sane to run commands just on one cloud.

emonty committed:
Use global inventory for single host

Because we add some hostvars and whatnot in the server list, and
because we loop all the servers for the single case anyway, just
use the list codepath so that we don't have to duplicate additional
keys.

emonty committed:
Rearrange into class in prep for work on caching

emonty committed:
Add caching for nova inventory

emonty committed:
Update copyright information to include HP

emonty committed:
Provide backwards compatibility for api_key option

emonty commented:
This includes and builds off of @CaptTofu work in #8388 I've tested that this works on both HP and Rackspace clouds - and that the multi-cloud support works with both of those clouds in the same config. I've also tested that it works both with and without config files, and that the config file overrides values found in the environment.

mpdehaan commented:
Looks reasonable to me, can you post to ansible-devel and see if we can get some +1s from some other openstack users perhaps?

My only other comment was some fields now have regions set in the config, if it worked for all regions before we _may_ want to see that region parameter commented out and have a default of all regions, or does that not make sense?

Thinking about people who upgrade that have a previous INI file, though since inventory plugins are not really part of a release, no one should really upgrade and get any surprises.

CaptTofu commented:
Michael,

Indeed, thank you. I emailed the dev list asking for reviewers. I think that my work and then Monty’s great work on top (I merged his pull request) will make this plugin top notch!

Regards,

Patrick

On Aug 18, 2014, at 1:56 PM, Michael DeHaan notifications@github.com wrote:

> Looks reasonable to me, can you post to ansible-devel and see if we can get some +1s from some other openstack users perhaps?
> 
> My only other comment was some fields now have regions set in the config, if it worked for all regions before we may want to see that region parameter commented out and have a default of all regions, or does that not make sense?
> 
> Thinking about people who upgrade that have a previous INI file, though since inventory plugins are not really part of a release, no one should really upgrade and get any surprises.
> 
> —
> Reply to this email directly or view it on GitHub.

emonty commented:
@mpdehaan - region is actually a required parameter for python-novaclient. The previous version filled in a fairly bogus region param if you didn't specify one - so I think it's fairly unlikely that someone was relying on that for a production usage - unless they actually have a production cloud with some very strange config ... I'll follow up on the mailing list thread.

ibratoev commented on plugins/inventory/nova.py:
```diff
@@ -30,103 +37,316 @@
+                if image_name:
+                    groups['image-%s' % image_name].append(server.name)
+
+                # And finally, add an IP address
```
This is duplicated 15 lines above that ;)

ibratoev commented on plugins/inventory/nova.py:
```diff
@@ -30,103 +37,316 @@
+
+                # And finally, add an IP address
+                if (self.private is True):
+                    hostvars[server.name]['ansible_ssh_host'] = private[0]
```
This will break the whole inventory if a single host has no network (for whatever reason).
Maybe ignore the host if it has no IP address?

CaptTofu replied:
Thank you very much for your review of this! Will fix.

emonty commented:
Oops on the duplication. Great catch on the breakage. Patch incoming...

emonty committed:
Ignore nova hosts with no network

If a host does not have working network, we should just skip it.

ibratoev commented:
What do you think about moving the _private_ option to the ini config file?
Currently, for one reason or another, I need to use _private_ ip addresses exclusively and to accomplish that I just changed the code itself.
Other than that the inventory is a great improvement - faster and cleaner than the current one. I have been using it extensively the last couple of weeks.
+1 for merging it sooner.

emonty commented:
@ibratoev I think that's a great idea! I'll do that right now.

emonty committed:
Allow setting private flag in nova.ini

It's useful for people at times to express that they always need to
connect to the private, not the public, interface.

emonty commented:
@ibratoev There you go - I tested this locally that it still works on the command line and that it also works setting it in the config file. Can you let me know if it works for you how you expect?

t0mk commented:
Hi, I did another rework of the Nova inventory, I tried to cover your features and add some more, and clean up the code. If you are interested, please see:
https://github.com/ansible/ansible/pull/9378
